src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java

```diff
@@ -856,7 +958,149 @@ public void configureMockConnection(final Connection connection) {
+    }
+
+    /**
+     * Invoked to remove from the known ad-hoc channels on a connection.
+     *
+     * @param resourceLocations The resource locations to remove.
+     * @param connection        The connection to remove the channels from.
+     */
+    private void onMinecraftUnregister(Set<ResourceLocation> resourceLocations, Connection connection) {
+        getKnownAdHocChannelsOfOtherEnd(connection).removeAll(resourceLocations);
+    }
+
+    /**
+     * {@return the initial channels that the server listens on during the configuration phase.}
+     */
+    public Set<ResourceLocation> getInitialServerListeningChannels() {
+        return Set.of(
+                MinecraftRegisterPayload.ID,
+                MinecraftUnregisterPayload.ID,
+                ModdedNetworkQueryPayload.ID);
+    }
+
+    public Set<ResourceLocation> getInitialServerUnregisterChannels() {
+        final ImmutableSet.Builder<ResourceLocation> nowForgottenChannels = ImmutableSet.builder();
+        nowForgottenChannels.add(MinecraftRegisterPayload.ID);
+        nowForgottenChannels.add(MinecraftUnregisterPayload.ID);
+        knownPlayRegistrations.entrySet().stream()
+                .filter(registration -> registration.getValue().flow().isEmpty() || registration.getValue().flow().get() == PacketFlow.SERVERBOUND)
+                .filter(registration -> registration.getValue().optional())
+                .forEach(registration -> nowForgottenChannels.add(registration.getKey()));
+        return nowForgottenChannels.build();
+    }
+
+    private static Set<ResourceLocation> getInitialClientListeningChannels() {
+        return Set.of(
+                MinecraftRegisterPayload.ID,
+                MinecraftUnregisterPayload.ID,
+                ModdedNetworkQueryPayload.ID,
+                ModdedNetworkSetupFailedPayload.ID,
+                ModdedNetworkPayload.ID);
+    }
+
+    public void onConfigurationFinished(ServerConfigurationPacketListener serverConfigurationPacketListener) {
+        final NetworkPayloadSetup setup = serverConfigurationPacketListener.getConnection().channel().attr(ATTRIBUTE_PAYLOAD_SETUP).get();
+        if (setup == null) {
+            LOGGER.error("Somebody tried to finish the configuration phase of a connection that has not negotiated with the client. Not finishing configuration.");
+            return;
+        }
+
+        final ImmutableSet.Builder<ResourceLocation> notListeningAnymoreOn = ImmutableSet.builder();
+        notListeningAnymoreOn.addAll(getInitialServerListeningChannels());
+        notListeningAnymoreOn.addAll(setup.configuration().keySet());
+        serverConfigurationPacketListener.send(new MinecraftUnregisterPayload(notListeningAnymoreOn.build()));
+
+        final ImmutableSet.Builder<ResourceLocation> nowListeningOn = ImmutableSet.builder();
+        nowListeningOn.add(MinecraftRegisterPayload.ID);
+        nowListeningOn.add(MinecraftUnregisterPayload.ID);
+        if (!setup.vanilla()) {
```

Comment:
The `vanilla` boolean should be removed from the `NetworkPayloadSetup`. This information is already accessible via the connection and/or the listener.